rollup/core/classes/RollupAsyncProcessor.cls

```diff
@@ -473,36 +463,59 @@ global virtual without sharing class RollupAsyncProcessor extends Rollup impleme
       return;
     }
 
-    Set<String> lookupKeys = lookupToCalcItems.keySet();
-    String whereClause = '' + rollup.lookupFieldOnCalcItem + ' = :lookupKeys';
     Map<Schema.SObjectType, Set<String>> localCalcToUniqueFieldNames = rollup.fullRecalcProcessor != null &&
       rollup.calcObjectToUniqueFieldNames != null
       ? rollup.calcObjectToUniqueFieldNames
       : this.calcObjectToUniqueFieldNames;
 
-    String countQuery = RollupQueryBuilder.Current.getQuery(rollup.calcItemType, new List<String>{ 'Count()' }, 'Id', '!=', whereClause);
-
-    // a dark hack, possibly even evil - but there are tests that rely on only a single record being batched at a time
-    // and we can't both consume the one query allowed AND do the countQuery call here. This is because actually deferring rollups
-    // isn't allowed in tests - we can't rebatch or requeue
-    Integer outstandingItemCount = Test.isRunningTest() ? 0 : Database.countQuery(countQuery);
-    Integer remainingQueryRowsLeft = Limits.getLimitQueryRows() - (Limits.getQueryRows() + 1);
-    shouldLimitCalcItemQuery = outstandingItemCount >= remainingQueryRowsLeft;
-
+    List<SObject> additionalCalcItems;
+    String whereClause = '' + rollup.lookupFieldOnCalcItem + ' = :lookupKeys';
     String query = RollupQueryBuilder.Current.getQuery(
       rollup.calcItemType,
       new List<String>(localCalcToUniqueFieldNames.get(rollup.calcItemType)),
       'Id',
       '!=',
       whereClause
     );
-    if (shouldLimitCalcItemQuery) {
-      query = RollupQueryBuilder.Current.getAllRowSafeQuery(rollup.calcItemType, query += '\nLIMIT ' + remainingQueryRowsLeft);
-    }
-    RollupLogger.Instance.log('gathering additional calc items with query:', query, LoggingLevel.FINE);
-    additionalCalcItems = Database.query(query);
-    if (shouldLimitCalcItemQuery == false) {
-      localCachedFullRecalcs.put(key, additionalCalcItems);
+
+    Boolean hasAlreadyBeenQueried = this.cachedQueryToAdditionalCalcItems.containsKey(query);
```

Comment:
Is there any risk of the cached calc items being out of date? Like if the calc items are also being updated in the same transaction as rollup doing its work, could the previously queried calc items have old data? Or does `RollupCalcItemReplacer` handle that scenario?

Reply:
the "same calc items, different values due to updates" scenario is handled through `RollupEvaluator`! Great question though

Reply:
🥳